Added Proposal (#408)
observatorium/observatorium: docs/proposals/accepted/observatorium.md

---
title: "2021-05: Observatorium"
menu:
  docs:
    parent: "acc-proposals"
toc: true
---

Taken from [Open Source Design Doc Template](https://docs.google.com/document/d/1zeElxolajNyGUB8J6aDXwxngHynh4iOuEzy3ylLc72U/edit#).

## Observatorium

Initial Observatorium Design Proposal

* **Owners:**

>   * `@bwplotka`

Reviewer: I have to say this is little bit disappointing for the proposal for "observatorium.md" as it disingenuously implies the idea for all of Observatorium is owned by you. Maybe you meant to be the owner for a refactor or of a new API? Otherwise should the whole team be here?

>   * `@bwplotka`

Reviewer: I don't think this is appropriate for the proposal for "observatorium.md" as it somewhat implies the idea for all of Observatorium is entirely owned by you. Maybe you meant to be the owner for a refactor or of a new API? Otherwise should the whole team be here?

Author: Well.. it's really about ... proposal? ;p

Author: but ok, I am fine with removing myself, it's def not about me

### Pitfalls of the current solution

All the currently available alternatives and reasons why such Observability design might be a better idea are stated in [Alternatives Section](#Alternatives).

## Goals

**Target Audience**: Engineers who want to provide observability services for monitoring, debugging and SLOs assurance for themselves, their teams or even SaaS customers.

An Observability System that has the following characteristics:

* Open Source first
  * Apache v2 license
  * Free to use
  * All (in-scope) contributions welcome.
* Ability to scrape/ingest (write), store, query (read), and retrieve metrics, logs and traces from multiple clusters and tenants. APIs for now:
  * Metrics
    * Write:
      * OpenMetrics (pull model)
      * Prometheus Remote Write API (push model)
    * Read:
      * Compatible with the Prometheus query APIs (e.g PromQL) making it possible to use any dashboard solution (e.g. Grafana).
  * Logs: Kubernetes and Pod logs (Loki API)
* Support recording rules and alerting based on data from multiple clusters and tenants.

>   * Allow users to configure alerts and recording rules with immediate effect (minutes)

Reviewer 2: Jaeger is creating an API that would broker requests from Jaeger Query clients to metrics backends. The UI, or other Query clients, could then query Jaeger for a narrow set of metrics for specific services/operations.

> * Unified installation and deployment model for Kubernetes and OpenShift using Operator, OLM, non Operator, helm and Openshift Templates. 

Reviewer 3: I'm not convinced `Unified installation and deployment model` is or should be a design goal of Observatorium.

Observatorium will generate a set of 'best practice' Kubernetes YAMLs, but IMO should not opine on how you install them in your cluster. That's a very different problem.

> * Unified installation and deployment model for Kubernetes and OpenShift using Operator, OLM, non Operator, helm and Openshift Templates. 

Author: Rephrased to `Opinionated installation and deployment model for Kubernetes using Operator, OLM and jsonnet templates. `

> * Support multi-tenancy for all APIs allowing to choose between soft and hard tenancy.

Reviewer 3: If Observatorium is an opinionated SaaS, should we not mandate one of {soft,hard} tenancy? (probably soft) That way we limit the number of options we need to consider when designing & implementing features.

Author: Yea, but our opinion is to use hard a lot, so we need both IMO (:

  * Multiple tenants can access the same metrics (cross tenancy)

> * Support centralized and in-cluster deployments (or mix) for all APIs. In particular, this means the following topologies:

Reviewer 3: This can be split out into two requirements:
* Support a mixture of topologies
* Ability to enable or disable different APIs

Author: Yea, I will remove it, I don't think we are supporting anything on client side.

Author: anything that comes before push

  * Single Cluster (everything runs inside the cluster).
  * Mix of in cluster monitoring/alerting and subset of/all metrics pushed to a central location.
  * The purely centralized push model
* Horizontally Scalable: supporting smaller as well as larger scale.
  * It should be possible to deploy the solution in small environments such as single-node clusters.
  * Scaling up is a matter of adding more instances rather than throwing more resources at the same instance.

> * Relatively pp, easy to operate, and ubiquitous storage even for long retention:

Reviewer 3: [pianissimo (very quiet)](https://www.google.com/search?q=pianissimo+symbol&ei=-nrcYLxp0oPyAvLGnogH&oq=pianissimo+symbol&gs_lcp=Cgdnd3Mtd2l6EAMyAggAMgIIADICCAAyAggAMgIIADIGCAAQFhAeOgcIABBHELADOgcIABCwAxBDOgoILhCwAxDIAxBDOgQIABBDOgIILjoICC4QxwEQrwFKBQg4EgExSgQIQRgAUOQLWPISYNsUaAFwAngAgAFkiAG3BJIBAzcuMZgBAKABAaoBB2d3cy13aXrIAQzAAQE&sclient=gws-wiz&ved=0ahUKEwi8zuWbxL_xAhXSgVwKHXKjB3EQ4dUDCA4&uact=5) :trollface:

> * Relatively pp, easy to operate, and ubiquitous storage even for long retention:

Author: fixed

  * Object storage is the main storage solution for longer retention.
  * Block storage is only for short term data or local buffers.

> * Ability to define data retention policies.

Author: clarfified

* Control over the granularity of the data access.
  * Allowing a tenant to read a subset of data.
  * Allowing a tenant to send a subset of the data off the cluster.
* Secure by default
  * All external-facing APIs support TLS for encryption
  * All external-facing APIs support OpenID for authentication.
  * All external-facing APIs support [OPA](https://www.openpolicyagent.org/) for authorization.
* High availability and solid failover handling.
  * For example, for the write path, the client side can survive the unavailability of a central cluster for e.g 30m without data loss.
* Built-in mechanisms to support production setups out of the box.
  * Rate-limiting.
  * Retries.
  * Sane defaults for metrics ingestion (remote write configuration, sharding, queues).

> * Ability to maintain alert routing (e.g Alertmanager).

Reviewer 4: Is Alertmanager also multi-tenant in observatorium? Is it configurable for users?

Reviewer 3: Not at present - though it is one of the obvious big features that users will ask for.

Author: Moving t to soft goals

  * Alerts API:
    * Alertmanager v2 API
  
## Soft Goals

Currently not in scope due to time constraints, but possible if you want to contribute:

* The ability for users to manage dashboards (e.g Grafana)
* Provide a single access point with access to all signals
  *  Discovery
* Support per tenant API quota and usage reports (e.g. show-back, billing).
* Support cold and hot storage.
* Ability to ingest (write), store, query (read), and visualize traces.
* Ability to ingest (write), store, query (read), and visualize profiles.
* Standard out of box alerts and monitoring rules
  * For the components that Observatorium will deploy

>   * For Kubernetes

Author: Good point

* Support Blackbox Probes.
* Support [OLAP](https://en.wikipedia.org/wiki/Online_analytical_processing) Analytics Queries.
* Ability to correlate signals and quickly navigate from one to another.
* Ability to ingest metrics via pure push method (push gateway, OLTP metrics protocol).
* Deploy the components on non-Kubernetes platforms.
* Pluggability: Supporting more than one project per signal

## Non-Goals

Currently not in scope due to project complexity, but we may revisit:

* Extensibility in Authn/Authz protocols.
* Define packaging and installation methods (this will be covered by a separate design doc)
* Non Observability Events.
* Other APIs then defined above.

## How

> TBD

Reviewer 3: Anything to add here? Feels like an important point ^^

Author: nah ;p

Author: that's trivial ;p

## Alternatives

"Here are the objections you should be having as you read this proposal. Here's why I still think we should take this path."

1. Don’t do it. Prometheus with Operator, Thanos, Loki, Jaeger, Tempo, Grafana exist, why not deploying and configuring on their own?

That’s a very valid point. You can already and those projects already collaborate with each other. However, to have a solid infrastructure and observability using those you need to spend many months of engineering time in understanding:

*   **What project to choose for each signal?**

>     Investing in a single project takes time and effort, so you want to choose properly. But how to choose? In Observatorium we chose for you based on Red Hat and open source community experience and the best of our knowledge.

Author: I don't think that conflicts, since it is a soft goal

*   **How to deploy it using a single way? **

    There is no single way. Each of the projects prefers different things. One prefers Ansible or Operator model more, second Jsonnet, other Helm. And you probably use something totally different. In most cases, your organization has to reinvent all of it and it takes time and huge knowledge to develop and maintain.

*   **Where to store data?**

Object Storage, Disk, Other DB?

*   **How to operate and run it?**

What are the alerting rules to deploy, what performance limits to set, what are the dashboards?

*   **How to have a unified API, rate limiting, quota, billing, and auth story?**

    Each project can use different APIs, protocols, and most likely enforce adding your own auth, rate limit, etc layers. Observatorium adds all of this.

> *   **Lot’s of data is duplicated. How to reduce running costs?**

Reviewer 3: ```suggestion
*   **Lots of data is duplicated. How can we reduce running costs?**
```

So in our opinion, the biggest value is in a system that gathers all metrics, logging, tracing, and profiles (if you choose to) and allows for a correlation between those.

3. Don’t do it. Paid, close source solutions exist.

There are many players aiming to solve similar goals, but in the close source: Tanzai, Splunk, DataDog, ElasticCloud, StackDriver (GCP), CloudWatch (AWS), AWS Prometheus Managed, Grafana Enterprise, Logz.io, Honeycomb.io (there are more, those are just examples). Why investing time in contributing, developing, and running Observatorium if there are good paid services?

There are a few reasons:

*   **Cost: **

All of the above are paid. Of course, running Observatorium is far from free. However, it allows you to make the decision if you want to run yourself or allow others to manage Observatorium as SaaS for you. (While currently, no SaaS option emerged, because we announced the project now, we expect many will emerge). Additionally, one of the Observatorium’s main goals is to limit operational (maintenance) and running costs to a minimum.

*   **Missing features or relying on one signal: **

> Most of the mentioned solutions focus only on part of the observability story. This is great, however, it might mean inconsistency and further complexity if you have to pay for logging to service X, use metrics with project Y, and pay to service Z to have traced. See

Reviewer 5: ```suggestion
Most of the mentioned solutions focus only on part of the observability story. This is great, however, it might mean inconsistency and further complexity if you have to pay for logging to service X, use metrics with project Y, and pay to service Z to have traces. See
```

> <p id="gdcalert9" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: undefined internal link (link text: "section above"). Did you generate a TOC? </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert10">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>

Reviewer 3: wat?

> [section above](#heading=h.dio3ps7jwoe5) why this is a bad thing.

Reviewer 3: google docs copy pasta
